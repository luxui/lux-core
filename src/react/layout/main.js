/**
 * @module react/layout/main
 * @memberof react/layout
 */

import React from 'react'; // `React` must be in scope when using JSX

import registry from '../../lib/componentRegistry';
import routing from '../../lib/routing';

import responseModelShape from '../responseModel.propType';

import '../error';
import '../rest/collection';
import '../rest/item';

function component(Component, props, error) {

  return (
    <main id="maincontent" role="main">
      <Component {...props.data} error={error || false} path={props.path} />
    </main>
  );
}
component.propTypes = responseModelShape;

function mainComponent(props) {
  if (props.error) {

    return component(registry('Lux.Error'), props, props.error);
  }

  try {
    const pageComponent = routing(props.path);

    if (pageComponent) {

      return component(pageComponent, props);
    }
  } catch (e) {

    return component(registry('Lux.Error'), props, e);
  }

  const type = (props.data.class || [])
    .filter(item => /^(?:collection|item)$/.test(item))[0];

  switch (type) {
    case 'collection':

      return component(registry('Lux.Rest.Collection'), props);
    case 'item':

      return component(registry('Lux.Rest.Item'), props);
    default:

      return component(registry('Lux.Error'), props,
        new Error(`No appropriate Component for "${props.path}".`));
  }
}

registry('Lux.Layout.Main', mainComponent, false);
