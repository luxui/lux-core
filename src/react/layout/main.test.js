jest.mock('../../lib/routing');
import routing from '../../lib/routing';

import React from 'react';
import renderer from 'react-test-renderer';

import registry from '../../lib/componentRegistry';

import './main';
const Collection = registry('Lux.Rest.Collection');
const ErrorComponent = registry('Lux.Lux.Error');
const Item = registry('Lux.Rest.Item');
const Main = registry('Lux.Layout.Main');

let response;

describe('Main (supplied Layout)', function () {
  afterEach(function () {
    registry('Lux.Rest.Collection', Collection);
    registry('Lux.Rest.Item', Item);
    registry('Lux.Error', ErrorComponent);
  });

  beforeEach(function () {
    response = {
      data: {
        actions: [],
        class: ['collection'],
        entities: [],
        links: [
          {
            href: '/',
            rel: ['self'],
            title: 'Something Savvy',
          }
        ],
        properties: {},
        title: 'Testing Main Component',
      },
      error: false,
      path: '/home',
      status: 200,
    };
  });

  it('should exist; and should be a function', function () {
    expect(typeof Main).toBe('function');
  });

  it('should render a page from the `routing` module', function () {
    routing.mockImplementationOnce(function mock(path) {
      if (path === response.path) {

        return () => (<pre>{JSON.stringify(response, null, ' ')}</pre>);
      }

      throw new Error('Implementation fail.');
    });

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Rest.Collection component', function () {
    registry('Lux.Rest.Collection', () => (<pre>{JSON.stringify(response, null, ' ')}</pre>));

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Rest.Item component', function () {
    response.data.class = ['item'];

    registry('Lux.Rest.Item', () => (<pre>{JSON.stringify(response, null, ' ')}</pre>));

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Error component; when the response is an error', function () {
    response.error = new Error('API response error.');

    registry('Lux.Error', () => (<pre>{JSON.stringify(response, null, ' ')}</pre>));

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Error component; when the class type is not: collection or item', function () {
    response.data.class = ['something-else'];

    registry('Lux.Error', () => (<pre>{JSON.stringify(response, null, ' ')}</pre>));

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the Error component; when the `class` property is absent', function () {
    delete response.data.class;

    registry('Lux.Error', () => (<pre>{JSON.stringify(response, null, ' ')}</pre>));

    const component = renderer.create(
      <Main {...response} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
