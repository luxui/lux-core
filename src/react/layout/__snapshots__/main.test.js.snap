// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main (supplied Layout) should render a page from the \`routing\` module 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "class": [
       "collection"
      ],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": false,
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;

exports[`Main (supplied Layout) should render the Error component; when the \`class\` property is absent 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": false,
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;

exports[`Main (supplied Layout) should render the Error component; when the class type is not: collection or item 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "class": [
       "something-else"
      ],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": false,
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;

exports[`Main (supplied Layout) should render the Error component; when the response is an error 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "class": [
       "collection"
      ],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": {},
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;

exports[`Main (supplied Layout) should render the Rest.Collection component 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "class": [
       "collection"
      ],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": false,
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;

exports[`Main (supplied Layout) should render the Rest.Item component 1`] = `
<main
  id="maincontent"
  role="main"
>
  <pre>
    {
     "data": {
      "actions": [],
      "class": [
       "item"
      ],
      "entities": [],
      "links": [
       {
        "href": "/",
        "rel": [
         "self"
        ],
        "title": "Something Savvy"
       }
      ],
      "properties": {},
      "title": "Testing Main Component"
     },
     "error": false,
     "path": "/home",
     "status": 200
    }
  </pre>
</main>
`;
